### PR TITLE
Add get and set APIs for the ZeRO-3 partitioned parameters

### PR DESCRIPTION
**Pull request by yiliu30**
The DeepSpeed currently supports a set of debugging APIs to [get](https://deepspeed.readthedocs.io/en/latest/zero3.html#debugging) and [set](https://deepspeed.readthedocs.io/en/latest/zero3.html#modifying-partitioned-states) the **full** model states (parameters, gradients, and optimizer states). However, in some scenarios, only **local states** are needed, for example, when pruning some model layers based on a local criterion. After calling `model_engine.step()`, we need to apply the local mask to the partitioned parameters owned by each process. Therefore, I am submitting this PR to introduce some new APIs for `get` and `set` ZeRO-3 partial model states.

### APIs intro
```python
def safe_get_local_fp32_param(param):
    """Get the fp32 partitioned parameter."""

def safe_get_local_grad(param):
    """Get the fp32 gradient of a partitioned parameter."""

def safe_get_local_optimizer_state(param, optim_state_key):
    """Get the fp32 optimizer state of a partitioned parameter."""

def safe_set_local_fp32_param(param, value):
    """Update the partitioned fp32 parameter."""

def safe_set_local_optimizer_state(param, value, optim_state_key):
    """Update the fp32 optimizer state of a partitioned parameter."""
```

### Usage
```python
# local API
from deepspeed.utils import (
    safe_get_local_fp32_param,
    safe_get_local_grad,
    safe_get_local_optimizer_state,
    safe_set_local_fp32_param,
    safe_set_local_optimizer_state
    )
```
### TODO
- [x] Add local APIs
- [x] Add UTs
- [x] Update Docs

@tjruwase

### CODE REVIEW

**Comment by yiliu30** [comment]
@microsoft-github-policy-service agree company="Intel"

**Comment by tjruwase** [comment]
@yiliu30, thanks for this great contribution. We currently don't have API for setting gradient, which is a bit trickier. Do you mind helping with that for both full and local scenarios? It is okay for that to be a separate PR. 

A PR for setting gradients will address these requests: 
1. #3310
2. https://github.com/microsoft/DeepSpeed/pull/4192#issuecomment-1703995117

**Comment by yiliu30** [comment]
> @yiliu30, thanks for this great contribution. We currently don't have API for setting gradient, which is a bit trickier. Do you mind helping with that for both full and local scenarios? It is okay for that to be a separate PR.
> 
> A PR for setting gradients will address these requests:
> 
> 1. [[REQUEST] How to access to the gradients (to manipulate) while the model is training? #3310](https://github.com/microsoft/DeepSpeed/issues/3310)
> 2. [Allow modification of zero partitioned parameters #4192 (comment)](https://github.com/microsoft/DeepSpeed/pull/4192#issuecomment-1703995117)

@tjruwase Sure, I'd be happy to support it with a follow-up PR.

**Comment by yiliu30** [comment]
@tjruwase, I updated the docs, could you please take a moment to review them, thx

File: deepspeed/utils/tensor_fragment.py [hunk 1]
```diff
@@ -185,6 +185,72 @@ def safe_get_full_grad(param):
...
+def safe_get_local_fp32_param(param):
+    """Get the fp32 partitioned parameter.
+        Args:
+            param (``torch.nn.Parameter``): A model parameter
+    """
+    # ZeRO stage 3 param
+    if hasattr(param, 'ds_id'):
+        return param._z3_optimizer.get_local_fp32_param(param)
+
+    return None
```
Comment (tjruwase) [inline deepspeed/utils/tensor_fragment.py:213]: What about support for zero stage 1/2 and bf16_optimizer?

File: deepspeed/utils/tensor_fragment.py [hunk 2]
```diff
@@ -185,6 +185,72 @@ def safe_get_full_grad(param):
...
+    """Get the fp32 optimizer state of a partitioned parameter.
+        Args:
+            param (``torch.nn.Parameter``): A model parameter
+            optim_state_key (``string``): Key value of optimizer state (e.g., `exp_avg` in Adam optimizer)
+    """
+    # ZeRO stage 3 param
+    if hasattr(param, 'ds_id'):
+        return param._z3_optimizer.get_local_fp32_param(param, optim_state_key)
+
+    return None
```
Comment (tjruwase) [inline deepspeed/utils/tensor_fragment.py:226]: What about support zero stage 1/2 and bf16_optimizer?

**Comment by tjruwase** [comment]
@yiliu30, I noticed that only zero-3 is supported for the local APIs. Is there any reason not to support other zero stages like the full APIs? Thanks

**Comment by topsy404** [comment]
> @yiliu30, I noticed that only zero-3 is supported for the local APIs. Is there any reason not to support other zero stages like the full APIs? Thanks

Hi @tjruwase, there is no special reason, just because we are working on some experiments with ZeRO-3.  I can add the support for zero 1 and 2 if needed.

**Comment by yiliu30** [comment]
> @yiliu30, I noticed that only zero-3 is supported for the local APIs. Is there any reason not to support other zero stages like the full APIs? Thanks

Hi @tjruwase, there is no special reason, just because we are working on some experiments with ZeRO-3. I can add the support for zero 1 and 2 if needed.

**Comment by tjruwase** [comment]
@yiliu30, thanks for the explanation. I think it is okay to add zero 1 & 2 support in a separate PR. I don't want to block your work. Your contribution is greatly appreciated.

**Review by tjruwase — APPROVED** [review]

**Comment by guoxiaoyu-gxy** [comment]
Hi, is there any plan or update to add "safe_set_full_grad" API? I need this api in my case for manipulate the gradients manually.

**Comment by tjruwase** [comment]
@guoxiaoyu-gxy, thanks for your interest. Unfortunately, we currently lack bandwidth to add this API.
